cnd/src/network/behaviour.rs

```diff
@@ -0,0 +1,160 @@
+use crate::network::protocol::{InboundProtocolConfig, Message, OutboundProtocolConfig};
+use libp2p::{
+    core::{ConnectedPoint, Multiaddr, PeerId},
+    swarm::{
+        NetworkBehaviour, NetworkBehaviourAction, OneShotHandler, PollParameters, ProtocolsHandler,
+    },
+};
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+};
+
+/// Network behaviour that handles the secret hash protocol.
+#[derive(Debug)]
+pub struct Behaviour {
+    /// Events that need to be yielded to the outside when polling.
+    events: VecDeque<NetworkBehaviourAction<OutboundProtocolConfig, OutEvent>>,
+}
+
+/// Event generated by the NetworkBehaviour and that the swarm will report back.
+#[derive(Clone, Copy, Debug)]
+pub enum OutEvent {
+    Rx(Message), // OutEvent containing secret hash message.
+    Sent,        // Empty/nil OutEvent i.e., `()`.
+}
+
+/// OutEvent when a peer sends us a message.
+impl From<Message> for OutEvent {
+    fn from(msg: Message) -> OutEvent {
+        OutEvent::Rx(msg)
+    }
+}
+
+/// OutEvent that occurs when we send a message.
+impl From<()> for OutEvent {
+    fn from(_: ()) -> Self {
+        OutEvent::Sent
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ProtocolsHandler = OneShotHandler<InboundProtocolConfig, OutboundProtocolConfig, OutEvent>;
+    type OutEvent = OutEvent;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        Default::default()
+    }
+
+    fn addresses_of_peer(&mut self, _: &PeerId) -> Vec<Multiaddr> {
+        Vec::new() // Announce protocol takes care of this.
+    }
+
+    fn inject_connected(&mut self, _: PeerId, _: ConnectedPoint) {
+        // Do nothing, announce protocol is going to take care of connections.
+    }
+
+    fn inject_disconnected(&mut self, _: &PeerId, _: ConnectedPoint) {
+        // Do nothing, announce protocol is going to take care of connections.
+    }
+
+    fn inject_node_event(&mut self, _: PeerId, event: OutEvent) {
+        // We ignore successful sends event.
+        let message = match event {
+            OutEvent::Rx(message) => message,
+            OutEvent::Sent => return,
+        };
+
+        // Add the message to be dispatched to the user.
+        self.events
+            .push_back(NetworkBehaviourAction::GenerateEvent(OutEvent::Rx(message)));
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut Context<'_>,
+        _: &mut impl PollParameters,
+    ) -> Poll<
+        NetworkBehaviourAction<
+            <Self::ProtocolsHandler as ProtocolsHandler>::InEvent,
+            Self::OutEvent,
+        >,
+    > {
+        if let Some(event) = self.events.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        Poll::Pending
```

Comment:
If you do this, you will have to get a reference to the current waker (`context.waker()`) and store it inside `self`.

Whenever you change the `events` queue (inside `inject_node_event`) you will then have to check if you currently have a waker `if let Some(waker) ...` and call `waker.wake()` to tell the runtime that you would like to be polled again :)

Reply:
The floodsub implementation does not use a waker?  I cut'n'pasta the `poll` and copied the ideas for the `inject_node_event`.  Are you sure we need it?